vispy/visuals/volume.py

```diff
@@ -246,12 +246,16 @@
     while (iter < nsteps) {{
         for (iter=iter; iter<nsteps; iter++)
         {{
-            // Get sample color
-            vec4 color = $sample(u_volumetex, loc);
-            float val = color.r;
-
-            {in_loop}
-
+            // Ignore this step if clipped out by the clipping planes
+            float is_shown = $clip_by_planes(loc);
+            if (is_shown >= 0)
+            {{
```

Comment:
I wonder if we can avoid having this code inside the loop, and instead modify the `start_loc` and `nsteps` of the ray.

Reply:
Mmm, I guess that could add a lot of complexity too. Alternatively, could we (or do we?) avoid this check if there are no clipping planes?

Reply:
That wasmy first thought as well, but then I thought that if the clipped volume is convex, it wouldn't work - which could happen with my original idea of having the ability to do both intersections and unions of clipped volumes :P. However, that's more complicated than I thought (you'd want to set up an expression to determine the order of unions and intersections...), so maybe moving this to the vertex shader is the way to go?

The check is currently not skipped if there are no clipping planes; the function simply returns 1.0.

Reply:
> if the clipped volume is concave, it wouldn't work 

Ah, I see. Interesting idea to do logic ops with the clipped volume!

> so maybe moving this to the vertex shader is the way to go?

You mean the logic for determining whether a loc is visible or not, right? I'm not in deep enough to see what makes it complicated. Would it make sense to do the complicated stuff in Python and generate the code for `clip_by_planes` accordingly?

Reply:
Yes, I mean that. I'm not sure either, I'm very new to glsl and shaders... I don't quite understand what the vertex shader does in the context of volumes, or how I would go from the `start_loc` to a `clipped_start_loc`.
We're limited on how much we can generate from python, because everything has to happen before compile time, so I'm not sure how much that can help.